File: geardataHUST/FMD.m
function Final_Mode = FMD(fs, x, FilterSize, CutNum, ModeNum, MaxIterNum)
    %
    %---------------
    % Input:
    %---------------
    %
    %   fs              :       sampling frequency of x
    %   x               :       signal to be analyzed  !!! x 要求为列信号！！！
    %   FilterSiz       :       filter size
    %   CutNum          :       the cut number of the whole frequency band
    %   ModeNum         :       the final mode number
    %   MaxIterNum      :       max iteration number
    %
    %---------------
    % Output:
    %---------------
    %
    %   Final_Mode      :       final mode
    %
    %---------------
    % Reference:
    %---------------
    %       Y. Miao, B. Zhang, C. Li, J. Lin, D. Zhang
    %       "Feature Mode Decomposition:New Decomposition Theory
    %       for Rotating Machinery Fault Diagnosis"
    %       IEEE Transactions on Industrial Electronics.2022
    %       DOI:10.1109/TIE.2022.3156156
    %-------------------------------------------------
    %
    % Author:  Boyao Zhang AND Yonghao Miao
    % Version: 2022/02
    %
    %-------------------------------------------------

    % Initialization
    freq_bound = 0:1 / CutNum:1 - 1 / CutNum;

    temp_filters = zeros(FilterSize, CutNum);
    for n = 1:length(freq_bound)
        w = window(@hanning,FilterSize);
        temp_filters(:, n) = fir1(FilterSize - 1, [freq_bound(n) + eps, freq_bound(n) + 1 / CutNum - eps], w)';
    end

    % result initialization
    result = cell(CutNum + 1, 5);
    result(1, :) = {'IterCount', 'Iterations', 'CorrMatrix', 'ComparedModeNum', 'StopNum'};

    % Iterations
    temp_sig = repmat(x, [1, CutNum]);

    itercount = 2;
    while 1
        iternum = 2;
        if itercount == 2
            iternum = MaxIterNum - (CutNum - ModeNum) * iternum;
        end

        result{itercount, 1} = iternum;

        for n = 1:size(temp_filters, 2)

            f_init = temp_filters(:, n);
            [y_Iter, f_Iter, k_Iter, T_Iter] = xxc_mckd(fs, temp_sig(:, n), f_init, iternum, [], 1, 0);
            result{itercount, 2}{n, 1} = y_Iter(:, end);
            result{itercount, 2}{n, 2} = f_Iter(:, end);
            result{itercount, 2}{n, 3} = k_Iter(:, end);
            result{itercount, 2}{n, 4} = abs(fft(f_Iter));
            result{itercount, 2}{n, 4} = result{itercount, 2}{n, 4}(1:FilterSize / 2, :);
            [~, index1] = max(result{itercount, 2}{n, 4});
            result{itercount, 2}{n, 5} = (index1 - 1) * (fs / FilterSize);
            result{itercount, 2}{n, 6} = T_Iter;

        end

        X = result{itercount, 2};

        for n = 1:size(X, 1)
            temp_sig(:, n) = X{n, 1}(:, end);
            temp_filters(:, n) = X{n, 2}(:, end);
        end

        CorrMatrix = abs(corrcoef(temp_sig));
        CorrMatrix = triu(CorrMatrix, 1);

        [I, J, ~] = max_IJ(CorrMatrix);
        Location = [I, J];

        XI = X{I, 1}(:, end);
        XJ = X{J, 1}(:, end);
        XI = XI - mean(XI);
        XJ = XJ - mean(XJ);

        T_1 = result{itercount, 2}{I, 6};
        KI = CK(XI, T_1, 1);
        T_2 = result{itercount, 2}{J, 6};
        KJ = CK(XJ, T_2, 1);

        if KI > KJ
            output = J;
        else
            output = I;
        end

        X(output, :) = [];
        temp_sig(:, output) = [];
        temp_filters(:, output) = [];

        result{itercount, 3} = CorrMatrix;
        result{itercount, 4} = Location;
        result{itercount, 5} = output;

        if size(temp_filters, 2) == ModeNum-1
            break
        end

        itercount = itercount + 1;

    end
    %
    for nn = 1:size(result{itercount, 2}, 1)
        Final_Mode(:, nn) = result{itercount, 2}{nn, 1}(:, end);
    end
end
function [y_final f_final ckIter T_final] = xxc_mckd(fs, x, f_init, termIter, T, M, plotMode)
    %---------------
    % Reference:
    %---------------
    %   Y. Miao, M. Zhao, J. Lin, Y. Lei
    %       "Application of an improved maximum correlated kurtosis deconvolution method
    %       for fault diagnosis of rolling element bearings"
    %       Mechanical Systems and Signal Processing, 92 (2017) 173 - 195.
    %   Y. Miao, M. Zhao, K. Liang, J. Lin
    %       "Application of an improved MCKDA for fault detection
    %       of wind turbine gear based on encoder signal"
    %       Renewable Energy, 151 (2020) 192 - 203.
    %-------------------------------------------------
    %
    % Author: @ Yonghao Miao
    %
    %-------------------------------------------------

    % Assign default values for inputs

    if (isempty(termIter))
        termIter = 30;
    end
    if (isempty(plotMode))
        plotMode = 0;
    end
    if (isempty(M))
        M = 3;
    end

    %%----------------------------
    if (isempty(T))
        xxenvelope = abs(hilbert(x)) - mean(abs(hilbert(x)));
        T = TT(xxenvelope, fs);
    end

    T = round(T);

    %%----------------------------
    x = x(:);
    L = length(f_init);
    N = length(x);

    % Calculate XmT
    XmT = zeros(L, N, M + 1);
    for (m = 0:M)
        for (l = 1:L)
            if (l == 1)
                XmT(l, (m * T + 1):end, m + 1) = x(1:N - m * T);
            else
                XmT(l, 2:end, m + 1) = XmT(l - 1, 1:end - 1, m + 1);
            end
        end
    end

    % Calculate the matrix inverse section
    Xinv = inv(XmT(:, :, 1) * XmT(:, :, 1)');

    f = f_init;
    ck_best = 0;

    % Initialize iteration matrices
    y = zeros(N, 1);
    b = zeros(L, 1);
    ckIter = [];

    n = 1;

    while n == 1 || (n <= termIter)
        % Compute output signal
        y = (f' * XmT(:, :, 1))';

        % Generate yt
        f_final(:, n) = f;
        yt = zeros(N, M);
        for (m = 0:M)
            if (m == 0)
                yt(:, m + 1) = y;
            else
                yt(T + 1:end, m + 1) = yt(1:end - T, m);
            end
        end

        % Calculate alpha
        alpha = zeros(N, M + 1);
        for (m = 0:M)
            alpha(:, m + 1) = (prod(yt(:, [1:m (m + 2):size(yt, 2)]), 2).^2) .* yt(:, m + 1);
        end

        % Calculate beta
        beta = prod(yt, 2);

        % Calculate the sum Xalpha term
        Xalpha = zeros(L, 1);
        for (m = 0:M)
            Xalpha = Xalpha + XmT(:, :, m + 1) * alpha(:, m + 1);
        end

        % Calculate the new filter coefficients
        f = sum(y.^2) / (2 * sum(beta.^2)) * Xinv * Xalpha;
        f = f / sqrt(sum(f.^2)); 
        % Calculate the ck value of this iteration
        ckIter(n) = sum(prod(yt, 2).^2) / (sum(y.^2)^(M + 1));
        % Update the best match filter if required
        if (ckIter(n) > ck_best)
            ck_best = ckIter(n);
        end

        %% Update the period
        %%----------------------------
        xyenvelope = abs(hilbert(y)) - mean(abs(hilbert(y)));
        T = TT(xyenvelope, fs);
        T = round(T);
        T_final = T;

        % Calculate XmT
        XmT = zeros(L, N, M + 1);
        for (m = 0:M)
            for (l = 1:L)
                if (l == 1)
                    XmT(l, (m * T + 1):end, m + 1) = x(1:N - m * T);
                else
                    XmT(l, 2:end, m + 1) = XmT(l - 1, 1:end - 1, m + 1);
                end
            end
        end

        % Calculate the matrix inverse section
        Xinv = inv(XmT(:, :, 1) * XmT(:, :, 1)');

        %%----------------------------
        y_final(:, n) = filter(f_final(:, n), 1, x);
        n = n + 1;
    end
end
function [T] = TT (y, fs)
    % estimate the period in y based on auto-correlation function.
    %---------------
    % Input:
    %---------------
    %
    %   y       :       signal to be analyzed
    %   fs      :       sampling frequency
    %
    %---------------
    % Output:
    %---------------
    %
    %   T       :       estimated period in sample
    %
    %-------------------------------------------------

    % find the maximum lag M
    M = fs;

    NA = xcorr(y, y, M, 'coeff');
    NA = NA(ceil(length(NA) / 2):end);

    % find first zero-crossing
    sample1 = NA(1);

    for lag = 2:length(NA)
        sample2 = NA(lag);

        if ((sample1 > 0) && (sample2 < 0))
            zeroposi = lag;
            break;
        elseif ((sample1 == 0) || (sample2 == 0))
            zeroposi = lag;
            break;
        else
            sample1 = sample2;
        end

    end

    % Cut from the first zero-crossing
    NA = NA(zeroposi:end);
    % Find the max position (time lag)
    % corresponding the max aside from the first zero-crossing
    [~, max_position] = max(NA);
    % Give the extimated period by autocorrelation
    T = zeroposi + max_position;

end

function ck = CK(x, T, M)

    % Calculate the Correlated Kurtosis

    if nargin < 3
        M = 2;
    end

    % Insure x is a row vector
    x = x(:)';

    N = length(x);
    x_shift = zeros(M + 1, N);

    x_shift(1, :) = x;

    for m = 1:M
        x_shift(m + 1, T + 1:end) = x_shift(m, 1:end - T);
    end

    ck = sum(prod(x_shift).^2) / sum(x.^2)^(M + 1);

end

function [I, J, M] = max_IJ(X)
    % Returns the row and column indices of the maximum in matrix X.

    [temp, tempI] = max(X);
    [M, J] = max(temp);
    I = tempI(J);
end

File: 03-FMD/python-script/FMD.py
import numpy as np
from scipy.signal import firwin, hann, hilbert, lfilter, correlate
from numpy.linalg import inv

from xxc_mckd import xxc_mckd
from utils import TT, CK, max_IJ


def FMD(fs, x, FilterSize, CutNum, ModeNum, MaxIterNum):
    x = np.array(x).reshape(-1, 1)
    freq_bound = np.arange(0, 1, 1 / CutNum) # length=7

    # 构造滤波器
    temp_filters = np.zeros((FilterSize, CutNum)) # 一列是一个filter 矩阵 FilterSize x CutNum
    for i, fb in enumerate(freq_bound):
        # hann() 生成一个 长度为 FilterSize 的 Hanning 窗
            # 返回值 w 是一个 一维 NumPy 数组，元素值在 0~1 之间. 
            # 数组 w 就是窗函数，在滤波或频谱分析时用来加权信号，减小频谱泄漏或滤波器的旁瓣。
        # firwin() 是 SciPy 里的 FIR 滤波器设计函数（Window Method）：  (原matlab代码中也是用的工具箱)
            # 返回一个 FIR 滤波器系数数组，长度 = FilterSize
            # window=w → 使用你之前定义的 Hanning 窗函数
            # pass_zero=False → 设计 带通或带阻滤波器，而不是低通
            # cutoff（这里 [fb + 1e-6, fb + 1/CutNum - 1e-6]） → 截止频率或频带
        # w = hann(FilterSize) 
        # temp_filters[:, i] = firwin(FilterSize, 
        #                             [fb + 1e-6, fb + 1/CutNum - 1e-6],
        #                             window=w, pass_zero=False)

        temp_filters[:, i] = firwin(FilterSize,
                            [fb + 1e-6, fb + 1/CutNum - 1e-6],  # 生成具有截止频率区间[fl,fu]的个 FIR 滤波器
                            window='hann', pass_zero=False)  # 每次初始化一列
        

        # np.tile 的作用
            # np.tile(A, reps) 用来 重复数组 A，生成更大的数组
            # reps 是重复次数，可以是整数或元组，表示每个维度的重复次数
    temp_sig = np.tile(x, (1, CutNum)) #原始信号重复7次
    itercount = 2
    result = {}

    while True:
        iternum = 2 if itercount != 2 else MaxIterNum - (CutNum - ModeNum) * 2 #
        result[itercount] = {"Iter": [], "Signal": [], "Filter": [], "Freq": [], "T": []}
        # result(1, :) = {'IterCount', 'Iterations', 'CorrMatrix', 'ComparedModeNum', 'StopNum'};
        for n in range(temp_filters.shape[1]):
            f_init = temp_filters[:, n]
            y_iter, f_iter, k_iter, T_iter = xxc_mckd(fs, temp_sig[:, n], f_init, iternum, None, 1, 0)

            result[itercount]["Signal"].append(y_iter[:, -1])
            result[itercount]["Filter"].append(f_iter[:, -1])
            result[itercount]["T"].append(T_iter)

        # 更新信号
        for n in range(len(result[itercount]["Signal"])):
            temp_sig[:, n] = result[itercount]["Signal"][n]
            temp_filters[:, n] = result[itercount]["Filter"][n]

        CorrMatrix = np.abs(np.corrcoef(temp_sig.T))
        CorrMatrix = np.triu(CorrMatrix, 1)

        I, J, _ = max_IJ(CorrMatrix)
        XI = temp_sig[:, I] - np.mean(temp_sig[:, I])
        XJ = temp_sig[:, J] - np.mean(temp_sig[:, J])

        KI = CK(XI, result[itercount]["T"][I], 1)
        KJ = CK(XJ, result[itercount]["T"][J], 1)
        output = J if KI > KJ else I  # 剔除相关性大的一组

        temp_sig = np.delete(temp_sig, output, axis=1)
        temp_filters = np.delete(temp_filters, output, axis=1)

        if temp_filters.shape[1] == ModeNum - 1:
            break

        itercount += 1

    Final_Mode = np.column_stack(result[itercount]["Signal"])
    return Final_Mode



if __name__ == "__main__":
    fs = 20000
    x = np.random.randn(6000)
    # %   fs              :       sampling frequency of x
    # %   x               :       signal to be analyzed
    # %   FilterSiz       :       filter size
    # %   CutNum          :       the cut number of the whole frequency band
    # %   ModeNum         :       the final mode number
    # %   MaxIterNum      :       max iteration number
    #============ FMD 参数 ============#
    filtersize = 30
    cutnum = 7
    modenum = 2
    maxiternum = 20

    mode = FMD(fs, x, filtersize, cutnum, modenum, maxiternum)
    print(mode.shape)